### ISSUE TITLE
Lualine replaces statusline created by vim-visual-multi

### ISSUE THREAD

**Issue by farzadmf** [nvim-lualine/lualine.nvim#951]
### Self Checks

- [x] I'm using the latest lualine.
- [x] I didn't find the issue in existing issues or PRs.

### How to reproduce the problem

1. Start with the minimal config below
2. Open a new buffer and have a couple of lines, something like:
```
hello world
I say hello
```
3. Go to `hello` in either line and press `Ctrl-n` (default keymap to start visual-multi mode)
4. VM (visual multi)'s update to statusline doesn't appear, or randomly appears and disappears

### Expected behaviour

When using [vim-airline](https://github.com/vim-airline/vim-airline), VM (visual multi), when enabled, seems to "take over" the status line, and it creates a status line with information about multi-cursors, which is updated as multi-cursors move around etc.

### Actual behaviour

Seems like lualine doesn't let VM have the statusline and keeps clearing it and redrawing it.

### Minimal config to reproduce the issue

```
call plug#begin("/private/tmp/algo/lualine/.local/share/nvim/plugged")

Plug 'nvim-lualine/lualine.nvim'
Plug 'https://github.com/mg979/vim-visual-multi'

call plug#end()

lua << END

require'lualine'.setup {
}
END
```

**Comment by jamestansx** [comment]
In the meantime, you could manually hide/unhide `lualine` with `autocmd`.

<details>
<summary>Example</summary>

```lua
vim.api.nvim_create_autocmd({ 'User' }, {
  pattern = 'visual_multi_start',
  callback = function()
    require('lualine').hide()
  end
})

vim.api.nvim_create_autocmd({ 'User' }, {
  pattern = 'visual_multi_exit',
  callback = function()
    require('lualine').hide({unhide=true})
  end
})
```

</details>

Ref: [VM-faq](https://github.com/mg979/vim-visual-multi/blob/724bd53adfbaf32e129b001658b45d4c5c29ca1a/doc/vm-faq.txt#L50)

**Comment by jamestansx** [comment]
I ain't sure why, could you try to reproduce with the following minimal config?

```lua
-- repro.lua
local root = vim.fn.fnamemodify("./.repro", ":p")
for _, name in ipairs({ "config", "data", "state", "cache" }) do
        vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
        vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath })
end
vim.opt.runtimepath:prepend(lazypath)

vim.opt.termguicolors = true

-- install plugins
local plugins = {
  {
    'nvim-lualine/lualine.nvim',
    config = true,
  },
  {
    'mg979/vim-visual-multi',
    version = false,
  },
}

vim.api.nvim_create_autocmd({ 'User' }, {
  pattern = 'visual_multi_start',
  callback = function()
    require('lualine').hide()
  end
})

vim.api.nvim_create_autocmd({ 'User' }, {
  pattern = 'visual_multi_exit',
  callback = function()
    require('lualine').hide({unhide=true})
  end
})

require("lazy").setup(plugins, {
        root = root .. "/plugins",
})

vim.cmd [[colorscheme habamax]]
```

`nvim -u /path/to/repro.lua`

**Comment by farzadmf** [comment]
@BrunoMSantos that's something at least

But it's sad. When I start VM, I see its status for a split second before it's replaced by `lualine`.

The original comment provided a way to hide/unhide lualine with an `autocmd`, but `lualine` seems to take the whole status line with it when being hidden, so nothing else appears.

This is really sad as `VM` is a plugin a main 100000000 times a day and I can't live without it; it would be so nice if we could find a solution 🙁

**Comment by BrunoMSantos** [comment]
Have you tried what I posted? That _is_ the solution. And if you improve on it, feel free to share.

**Comment by farzadmf** [comment]
I did, but the "built-in" status from `VM` looks nicer; it has colors and everything, and, for some reason, it looks shorter than this solution, not sure why

But yeah, your solution works

**Comment by farzadmf** [comment]
You're right. I'm just too lazy to look into things; I just want something that "works" 😆

**Comment by musjj** [comment]
@jamestansx

> I ain't sure why, could you try to reproduce with the following minimal config?

Weirdly enough, I *can't* reproduce it in the minimal config. I'm not sure which of my plugins is causing this behavior. 

I tried using the default config for lualine, but the issue still happens in my config. I just don't have a clue on what could possible be interfering.

My config has grown pretty hairy recently, so the good-old halving till it works is kinda hard to do.

**Comment by nnhutan** [comment]
```lua
local function get_visual_multi()
	local result = vim.fn["VMInfos"]()
	-- local current = result.current
	-- local total = result.total
	local ratio = result.ratio
	local patterns = result.patterns
	-- local status = result.status
	return "%#St_InsertMode# "
		.. " MULTI "
		.. "%#St_lspWarning#  "
		.. patterns[1]
		.. " "
		.. "%#StText#"
		.. " "
		.. ratio
end
```


Another approach to get useful information for statusline. We can use this to add to the custom section in the statusline

**Comment by BrunoMSantos** [comment]
> ```lua
> local function get_visual_multi()
> 	local result = vim.fn["VMInfos"]()
> 	-- local current = result.current
> 	-- local total = result.total
> 	local ratio = result.ratio
> 	local patterns = result.patterns
> 	-- local status = result.status
> 	return "%#St_InsertMode# "
> 		.. " MULTI "
> 		.. "%#St_lspWarning#  "
> 		.. patterns[1]
> 		.. " "
> 		.. "%#StText#"
> 		.. " "
> 		.. ratio
> end
> ```
> 
> Another approach to get useful information for statusline. We can use this to add to the custom section in the statusline

Thanks for sharing, managed to improve mine with this. One thing though, what are the '%#St_lspWarning#' bits about? Can't find that in documentation and don't see any difference between having those or not.

**Comment by nnhutan** [comment]
> > ```lua
> > local function get_visual_multi()
> > 	local result = vim.fn["VMInfos"]()
> > 	-- local current = result.current
> > 	-- local total = result.total
> > 	local ratio = result.ratio
> > 	local patterns = result.patterns
> > 	-- local status = result.status
> > 	return "%#St_InsertMode# "
> > 		.. " MULTI "
> > 		.. "%#St_lspWarning#  "
> > 		.. patterns[1]
> > 		.. " "
> > 		.. "%#StText#"
> > 		.. " "
> > 		.. ratio
> > end
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Another approach to get useful information for statusline. We can use this to add to the custom section in the statusline
> 
> Thanks for sharing, managed to improve mine with this. One thing though, what are the '%#St_lspWarning#' bits about? Can't find that in documentation and don't see any difference between having those or not.
Those are just the custom highlight groups for my custom statusline. You can arbitrarily use the information in the table returned when calling `vim.fn["VMInfos"]()`

**Comment by BrunoMSantos** [comment]
Ah, right. Yeah, don't think I want that, but thanks for the tips ;)

**Comment by webdevcat-me** [comment]
I wanted to have vim-visual-multi's `special_statusline` modes visible in the statusline whenever they were active, so here's what I ended up going with:

```lua
vim.g.VM_set_statusline = 0
vim.g.VM_silent_exit = 1

local function vm_mode()
  return vim.iter(string.gmatch(vim.fn['vm#themes#statusline'](), "%S+")):nth(2)
end

local function vm_status()
  return vim.fn['VMInfos']().status or ''
end

require('lualine').setup({
  sections = {
    lualine_a = {{'mode', fmt = function(mode) return vm_mode() or mode end}},
    lualine_b = {vm_status, 'branch', 'diff', 'diagnostics'},
  },
})

" A redraw is absolutely necessary to see the special statusline modes.
" And it has to be asynchronous. A synchronous call will not work,
" hence the use of `defer_fn()` with a delay of 0.
vim.api.nvim_create_autocmd({'CmdlineEnter'}, {
  pattern = {'@'},
  callback = function(ev)
    if vim.b.visual_multi then
      local delay = 0
      vim.defer_fn(function() vim.cmd('execute "redrawstatus"') end, delay)
    end
  end
})
```
Here's what it looks like in VM visual line mode:

![Screenshot from 2024-07-21 19-28-56](https://github.com/user-attachments/assets/ef4fbc3d-01f0-4bea-a42e-4361b3b4fc92)

And here it is in the special "Run visual" mode (`<Leader>+v` after switching to *extend mode* with `<Tab>`):

![Screenshot from 2024-07-21 19-30-17](https://github.com/user-attachments/assets/be92ff05-2d25-45fa-944f-10f7f018ce9e)

Without the redraw, the special modes would all just look like this:

![Screenshot from 2024-07-21 19-36-28](https://github.com/user-attachments/assets/72f75f77-7fd1-4172-9e65-ac8ca5d2e0a8)

And this is when VM isn't active:

![Screenshot from 2024-07-21 19-51-05](https://github.com/user-attachments/assets/7cc1c03d-ea34-4567-aea0-a73569eef31b)
